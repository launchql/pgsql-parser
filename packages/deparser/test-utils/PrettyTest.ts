import { expectParseDeparse } from './index';

const generateCoded = require('../../../__fixtures__/generated/generated.json');

export class PrettyTest {
  private testCases: string[];
  private testPrefix: string;

  constructor(testCases: string[], testPrefix: string = 'test') {
    this.testCases = testCases;
    this.testPrefix = testPrefix;
  }

  /**
   * Generate individual tests for each test case with both pretty and non-pretty formatting
   */
  generateTests(): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = `${this.testPrefix}-${index + 1}`;

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate tests with custom test name extraction
   */
  generateTestsWithCustomNaming(nameExtractor: (key: string, index: number) => string): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = nameExtractor(key, index);

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate a single AST validation test for all test cases
   */
  generateASTValidationTest(): void {
    it('should validate AST equivalence for all test cases', async () => {
      const allSql = this.testCases.map((key) => generateCoded[key]);
      
      for (const sql of allSql) {
        await expectParseDeparse(sql, { pretty: false });
        await expectParseDeparse(sql, { pretty: true });
      }
    });
  }
} 