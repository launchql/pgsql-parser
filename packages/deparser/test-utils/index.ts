import { parse } from '@pgsql/parser';
import { deparse } from '../src';
import { cleanTree, cleanLines } from '../src/utils';
import { readFileSync } from 'fs';
import * as path from 'path';
import { expect } from '@jest/globals';
import {diff} from 'jest-diff'

type ParseErrorType = 
  | 'PARSE_FAILED'
  | 'INVALID_STATEMENT'
  | 'REPARSE_FAILED'
  | 'AST_MISMATCH'
  | 'UNEXPECTED_ERROR'
  | 'INVALID_DEPARSED_SQL';

interface ParseError extends Error {
  type: ParseErrorType;
  relativePath: string;
  sql: string;
  deparsedSql?: string;
  originalAst?: any;
  reparsedAst?: any;
  parseError?: string;
}

function createParseError(
  type: ParseErrorType,
  relativePath: string,
  sql: string,
  deparsedSql?: string,
  originalAst?: any,
  reparsedAst?: any,
  parseError?: string
): ParseError {
  const error = new Error(getErrorMessage(type)) as ParseError;
  error.type = type;
  error.relativePath = relativePath;
  error.sql = sql;
  error.deparsedSql = deparsedSql;
  error.originalAst = originalAst;
  error.reparsedAst = reparsedAst;
  error.parseError = parseError;
  return error;
}

function getErrorMessage(type: ParseErrorType): string {
  switch (type) {
    case 'PARSE_FAILED':
      return 'Parse failed - no statements returned';
    case 'INVALID_STATEMENT':
      return 'Invalid statement structure';
    case 'REPARSE_FAILED':
      return 'Reparse failed - no statements returned';
    case 'AST_MISMATCH':
      return 'AST mismatch after parse/deparse cycle';
    case 'UNEXPECTED_ERROR':
      return 'Unexpected error during parse/deparse cycle';
    case 'INVALID_DEPARSED_SQL':
      return 'Invalid deparsed SQL';
  }
}

export class TestUtils {
  protected printErrorMessage(sql: string, position: number) {
    const lineNumber = sql.slice(0, position).match(/\n/g)?.length || 0;
    const lines = sql.split('\n');
    let colNumber = position - 1;
    for (let l = 0; l < lineNumber; l++) {
      colNumber -= lines[l].length + 1;
    }
    const errMessage = [`Error line ${lineNumber + 1}, column ${colNumber + 1}`];
    if (lineNumber > 0) {
      errMessage.push(lines[lineNumber - 1]);
    }
    errMessage.push(lines[lineNumber]);
    errMessage.push(' '.repeat(colNumber) + '^');
    if (lineNumber < lines.length - 1) {
      errMessage.push(lines[lineNumber + 1]);
    }
    console.error(errMessage.join('\n'));
  }

  tryParse(sql: string) {
    try {
      return parse(sql);
    } catch (err: any) {
      if (err.cursorPosition) {
        this.printErrorMessage(sql, err.cursorPosition);
      }
      throw err;
    }
  }

  expectAstMatch(relativePath: string, sql: string) {
    let tree: any;
    try {
      tree = this.tryParse(sql);
      if (tree.stmts) {
        tree.stmts.forEach((stmt: any) => {
          if (stmt.stmt) {
            const outSql = deparse(stmt.stmt);
            let reparsed;
            try {
              reparsed = parse(outSql);
            } catch (parseErr) {
              throw createParseError(
                'INVALID_DEPARSED_SQL',
                relativePath,
                sql,
                outSql,
                cleanTree([stmt]),
                undefined,
                parseErr instanceof Error ? parseErr.message : String(parseErr)
              );
            }
            const originalClean = cleanTree([stmt]);
            const reparsedClean = cleanTree(reparsed.stmts || []);
            
            if (!tree.stmts) {
              throw createParseError('PARSE_FAILED', relativePath, sql);
            }
            
            if (!stmt.stmt) {
              throw createParseError('INVALID_STATEMENT', relativePath, sql, undefined, cleanTree(tree));
            }
            
            if (!reparsed.stmts) {
              throw createParseError('REPARSE_FAILED', relativePath, sql, outSql, originalClean);
            }
            
            try {
              expect(reparsedClean).toEqual(originalClean);
            } catch (err) {
              throw createParseError('AST_MISMATCH', relativePath, sql, outSql, originalClean, reparsedClean);
            }
          }
        });
      }
    } catch (err) {
      const errorMessages: string[] = [];
      
      if (err instanceof Error && 'type' in err) {
        const parseError = err as ParseError;
        errorMessages.push(`\n❌ ${parseError.type}: ${parseError.relativePath}`);
        errorMessages.push(`❌ INPUT SQL: ${parseError.sql}`);
        
        if (parseError.deparsedSql) {
          errorMessages.push(`❌ DEPARSED SQL: ${parseError.deparsedSql}`);
        }
        
        if (parseError.type === 'INVALID_DEPARSED_SQL') {
          errorMessages.push(
            `\n❌ DEPARSER GENERATED INVALID SQL:`,
            `ORIGINAL AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nPARSE ERROR: ${parseError.parseError}`
          );
        } else if (parseError.type === 'AST_MISMATCH') {
          errorMessages.push(
            `\n❌ AST COMPARISON:`,
            `EXPECTED AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nACTUAL AST:`,
            JSON.stringify(parseError.reparsedAst, null, 2),
            `\nDIFF (what's missing from actual vs expected):`,
            diff(parseError.originalAst, parseError.reparsedAst)
          );
        } else if (parseError.originalAst) {
          errorMessages.push(`❌ AST: ${JSON.stringify(parseError.originalAst, null, 2)}`);
        }
      } else {
        // Handle unexpected errors
        errorMessages.push(
          `\n❌ UNEXPECTED ERROR: ${relativePath}`,
          `❌ INPUT SQL: ${sql}`,
          `❌ ERROR: ${err instanceof Error ? err.message : String(err)}`
        );
        if (tree) {
          errorMessages.push(`❌ PARSED AST: ${JSON.stringify(cleanTree(tree), null, 2)}`);
        }
      }
      
      console.log(errorMessages.join('\n'));
      throw err;
    }
  }
}

export class FixtureTestUtils extends TestUtils {
  private fixtures: Record<string, string>;

  constructor() {
    super();
    const GENERATED_JSON = path.join(__dirname, '../../../__fixtures__/generated/generated.json');
    this.fixtures = JSON.parse(readFileSync(GENERATED_JSON, 'utf-8'));
  }

  getTestEntries(filters: string[]) {
    if (filters.length === 0) {
      return Object.entries(this.fixtures);
    }
    
    return Object.entries(this.fixtures).filter(([relPath]) => 
      filters.includes(relPath)
    );
  }

  runFixtureTests(filters: string[]) {
    this.getTestEntries(filters).forEach(([relativePath, sql]) => {
      try {
        this.expectAstMatch(relativePath, sql);
      } catch (err) {
        throw err;
      }
    });
  }
}
