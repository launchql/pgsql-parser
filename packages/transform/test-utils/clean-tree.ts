/* eslint-disable no-restricted-syntax */

export const cleanLines = (sql: string) => {
    return sql
      .split('\n')
      .map((l) => l.trim())
      .filter((a) => a)
      .join('\n');
  };
  
  export const transform = (obj: any, props: any): any => {
    let copy: any = null;
    // Handle the 3 simple types, and null or undefined
    if (obj == null || typeof obj !== 'object') {
      return obj;
    }
  
    // Handle Date
    if (obj instanceof Date) {
      copy = new Date();
      copy.setTime(obj.getTime());
      return copy;
    }
  
    // Handle Array
    if (obj instanceof Array) {
      copy = [];
      for (let i = 0, len = obj.length; i < len; i++) {
        copy[i] = transform(obj[i], props);
      }
      return copy;
    }
  
    // Handle Object
    if (obj instanceof Object || typeof obj === 'object') {
      copy = {};
      for (const attr in obj) {
        if (obj.hasOwnProperty(attr)) {
          if (props.hasOwnProperty(attr)) {
            if (typeof props[attr] === 'function') {
              copy[attr] = props[attr](obj[attr]);
            } else if (props[attr].hasOwnProperty(obj[attr])) {
              copy[attr] = props[attr][obj[attr]];
            } else {
              copy[attr] = transform(obj[attr], props);
            }
          } else {
            copy[attr] = transform(obj[attr], props);
          }
        } else {
          copy[attr] = transform(obj[attr], props);
        }
      }
      return copy;
    }
  
    throw new Error("Unable to copy obj! Its type isn't supported.");
  };
  
  const noop = (): undefined => undefined;
  
  export const cleanTree = (tree: any) => {
    return transform(tree, {
      stmt_len: noop,
      stmt_location: noop,
      location: noop
    });
  };
  
  export const cleanTreeWithStmt = (tree: any) => {
    return transform(tree, {
      stmt_location: noop,
      location: noop
    });
  };
  