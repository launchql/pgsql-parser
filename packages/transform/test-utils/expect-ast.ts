import { cleanTree } from './clean-tree';

import { Parser } from '@pgsql/parser';

// TODO: These imports will be updated once the transform implementation is complete
// For now, we're creating the structure for the test utilities

export async function expectOldAstToTransformToNewAst(
  sql: string, 
  oldVersion: 13 | 14 | 15 | 16 | 17, 
  newVersion: 13 | 14 | 15 | 16 | 17,
  transform: (ast: any) => any
): Promise<void> {

    const parser = new Parser(oldVersion);
    const ast = await parser.parse(sql);

    const newParser = new Parser(newVersion);
    const newAst = await newParser.parse(sql);

    const transformedAst = transform(ast);

    // IMPORTANT: We need to clean the ASTs before comparing them
    // Otherwise, the location information will differ and the test will fail
    expect(cleanTree(transformedAst)).toEqual(cleanTree(newAst));
    
  // TODO: Implementation will be added once we have:
  // 1. A way to parse SQL with specific PostgreSQL versions
  // 2. The transformAST function implemented
  
  // This will:
  // - Parse SQL with the old parser version
  // - Transform the old AST to the new version
  // - Parse SQL with the new parser version for comparison
  // - Clean both ASTs to remove location information
  // - Compare the transformed AST with the natively parsed new AST
  
  throw new Error('Not implemented yet - waiting for parser with version support and transformAST implementation');
}

// Additional utility for testing with custom expectations
export async function expectTransformWithCustomAssertion(
  sql: string,
  oldVersion: number,
  newVersion: number,
  assertion: (transformedAst: any, newAst: any) => void
): Promise<void> {
  // TODO: Implementation will be added once we have parser and transform support
  throw new Error('Not implemented yet - waiting for parser with version support and transformAST implementation');
}

// Utility to test transformation with stats
export async function expectTransformWithStats(
  sql: string,
  oldVersion: number,
  newVersion: number
): Promise<{
  transformedAst: any;
  newAst: any;
  stats: any;
}> {
  // TODO: Implementation will be added once we have parser and transform support
  throw new Error('Not implemented yet - waiting for parser with version support and transformASTWithResult implementation');
}