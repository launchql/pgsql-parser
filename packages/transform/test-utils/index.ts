import { Parser } from '@pgsql/parser';
import { cleanTree } from './clean-tree';
import { readFileSync } from 'fs';
import * as path from 'path';
import { expect } from '@jest/globals';
import { diff } from 'jest-diff';
import { skipTests } from './skip-tests';
const parser13 = new Parser({ version: 13 });
const parser14 = new Parser({ version: 14 });
const parser15 = new Parser({ version: 15 });
const parser16 = new Parser({ version: 16 });
const parser17 = new Parser({ version: 17 });

import { ASTTransformer } from '../src/multi-version-transformer';

/**
 * Get the appropriate parsers for version transformation
 */
export function getParsersForVersions(versionPrevious: number, versionNext: number): { parserPrevious: any, parserNext: any } {
  if (versionNext - versionPrevious !== 1) {
    throw new Error('Version difference must be 1');
  }

  let parserPrevious;
  let parserNext;
  
  switch (versionPrevious) {
    case 13:
      parserPrevious = parser13;
      parserNext = parser14;
      break;
    case 14:
      parserPrevious = parser14;
      parserNext = parser15;
      break;
    case 15:
      parserPrevious = parser15;
      parserNext = parser16;
      break;
    case 16:
      parserPrevious = parser16;
      parserNext = parser17;
      break;
    default:
      throw new Error('Unsupported version');
  }

  return { parserPrevious, parserNext };
}

/**
 * Get the appropriate transformer for version transformation
 */
export function getTransformerForVersion(versionPrevious: number, versionNext: number): ASTTransformer {
  // Return a single ASTTransformer instance that can handle all transformations
  return new ASTTransformer();
}

/**
 * Perform the parse-transform-parse equality test
 */
export async function expectTransformedAstToEqualParsedAst(
  sql: string,
  parserPrevious: any,
  parserNext: any,
  transformer: ASTTransformer,
  versionPrevious: number,
  versionNext: number,
  relativePath?: string
): Promise<void> {
  const parsedPrevious = await parserPrevious.parse(sql);
  const parsedNext = await parserNext.parse(sql);
  
  // Create a copy of the previous AST to transform
  const astToTransform = JSON.parse(JSON.stringify(parsedPrevious));
  
  // Transform the statements within the AST
  if (astToTransform.stmts && Array.isArray(astToTransform.stmts)) {
    astToTransform.stmts = astToTransform.stmts.map((stmtWrapper: any, index: number) => {
      if (stmtWrapper.stmt) {
        try {
          // Transform the actual statement using the ASTTransformer
          const transformedStmt = transformer.transform(stmtWrapper.stmt, versionPrevious, versionNext);
          return { ...stmtWrapper, stmt: transformedStmt };
        } catch (error: any) {
          const errorMessage = [
            `\n❌ TRANSFORMATION ERROR ${relativePath ? `(${relativePath})` : ''}`,
            `   ⚠️ Previous Version: ${versionPrevious}`,
            `   ⚠️ Next Version: ${versionNext}`,
            `   ⚠️ Statement Index: ${index}`,
            `   ⚠️ Statement Type: ${Object.keys(stmtWrapper.stmt)[0]}`,
            `   ⚠️ Error: ${error.message}`,
            `\n   ⚠️ Original Statement:`,
            JSON.stringify(stmtWrapper.stmt, null, 2)
          ].join('\n');
          
          console.error(errorMessage);
          throw error;
        }
      }
      return stmtWrapper;
    });
  }
  
  // Update the version to match the target version
  astToTransform.version = parsedNext.version;
  
  const nextAst = cleanTree(parsedNext);
  const previousTransformedAst = cleanTree(astToTransform);
  
  // Remove version fields for comparison
  delete nextAst.version;
  delete previousTransformedAst.version;
  
  try {
    expect(nextAst).toEqual(previousTransformedAst);
  } catch (error: any) {
    // Try to find the first difference
    const d = diff(nextAst, previousTransformedAst);
    
    const errorMessage = [
      `\n❌ TRANSFORMATION MISMATCH ${relativePath ? `(${relativePath})` : ''}`,
      `   ⚠️ Previous Version: ${versionPrevious}`,
      `   ⚠️ Next Version: ${versionNext}`,
      `   ⚠️ SQL: ${sql}`,
      `\n   ⚠️ Expected (parsed with v${versionNext}):`,
      JSON.stringify(nextAst, null, 2),
      `\n   ⚠️ Actual (transformed from v${versionPrevious}):`,
      JSON.stringify(previousTransformedAst, null, 2)
    ].filter(line => line !== '').join('\n');
    console.log(relativePath + ':\n' + d);
    console.error(errorMessage);
    throw error;
  }
}

/**
 * Combined helper that uses all three modular functions
 */
export async function expectParseTransformParseToBeEqual(sql: string, versionPrevious: number, versionNext: number) {
  const { parserPrevious, parserNext } = getParsersForVersions(versionPrevious, versionNext);
  const transformer = getTransformerForVersion(versionPrevious, versionNext);
  
  await expectTransformedAstToEqualParsedAst(sql, parserPrevious, parserNext, transformer, versionPrevious, versionNext);
}

export class FixtureTestUtils {
  private fixtures: Record<string, string>;
  versionPrevious: number;
  versionNext: number;

  transformer: ASTTransformer;
  parserPrevious: any;
  parserNext: any;

  constructor(versionPrevious: number, versionNext: number) {
    if (versionNext - versionPrevious !== 1) {
      throw new Error('Version difference must be 1');
    }
    this.versionNext = versionNext;
    this.versionPrevious = versionPrevious;

    // Use the modular helper functions
    this.transformer = getTransformerForVersion(versionPrevious, versionNext);
    const { parserPrevious, parserNext } = getParsersForVersions(versionPrevious, versionNext);
    this.parserPrevious = parserPrevious;
    this.parserNext = parserNext;

    const GENERATED_JSON = path.join(__dirname, '../../../__fixtures__/generated/generated.json');
    this.fixtures = JSON.parse(readFileSync(GENERATED_JSON, 'utf-8'));
  }

  getTestEntries(filters: string[]) {
    if (filters.length === 0) {
      return Object.entries(this.fixtures);
    }
    return Object.entries(this.fixtures).filter(([relPath]) =>
      filters.includes(relPath)
    );
  }

  async expectParseTransformParseToBeEqual(relativePath: string, sql: string) {
    // Use the modular helper function instead of duplicating logic
    await expectTransformedAstToEqualParsedAst(sql, this.parserPrevious, this.parserNext, this.transformer, this.versionPrevious, this.versionNext, relativePath);
  }

  async runFixtureTests(filters: string[]) {
    if (filters.length === 0) {
      console.log('no filters provided, skipping tests.');
      return;
    }
    
    const entries = this.getTestEntries(filters);
    for (const [relativePath, sql] of entries) {
      // Check if this test should be skipped
      const skipEntry = skipTests.find(([versionPrevious, versionNext, test, reason]) => 
        versionPrevious === this.versionPrevious && 
        versionNext === this.versionNext && 
        test === relativePath
      );
      
      if (skipEntry) {
        console.log(`⏭️  SKIPPING: ${relativePath} - ${skipEntry[3]}`);
        continue;
      }
      
      try {
        await this.expectParseTransformParseToBeEqual(relativePath, sql);
      } catch (error: any) {
        console.error(`\n❌ FAILED: ${relativePath}`);
        throw error;
      }
    }
  }
}

// Re-export the full transform flow utilities
export { FullTransformFlowFixture } from './full-transform-flow';

/**
 * Legacy fixture utility for testing direct transformations
 * Uses the legacy SQL fixtures for testing direct transformers
 */
export class LegacyFixtureTestUtils {
  private fixtures: Record<string, string>;

  constructor() {
    const LEGACY_JSON = path.join(__dirname, '../../../__fixtures__/legacy/13-legacy-check.json');
    this.fixtures = JSON.parse(readFileSync(LEGACY_JSON, 'utf-8'));
  }

  getTestEntries(filters: string[]) {
    if (filters.length === 0) {
      return Object.entries(this.fixtures);
    }
    return Object.entries(this.fixtures).filter(([relPath]) =>
      filters.some(filter => relPath.includes(filter))
    );
  }

  getAllTestEntries() {
    return Object.entries(this.fixtures);
  }

  getTestEntry(key: string): string | undefined {
    return this.fixtures[key];
  }

  async runLegacyTests(filters: string[], testFn: (relativePath: string, sql: string) => Promise<void>) {
    if (filters.length === 0) {
      console.log('no filters provided, skipping tests.');
      return;
    }
    
    const entries = this.getTestEntries(filters);
    for (const [relativePath, sql] of entries) {
      try {
        await testFn(relativePath, sql);
      } catch (error: any) {
        console.error(`\n❌ FAILED: ${relativePath}`);
        throw error;
      }
    }
  }

  async runAllLegacyTests(testFn: (relativePath: string, sql: string) => Promise<void>) {
    const entries = this.getTestEntries([]);
    for (const [relativePath, sql] of entries) {
      try {
        await testFn(relativePath, sql);
      } catch (error: any) {
        console.error(`\n❌ FAILED: ${relativePath}`);
        throw error;
      }
    }
  }
}
