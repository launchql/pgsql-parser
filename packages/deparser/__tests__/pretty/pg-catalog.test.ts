import { PrettyTest } from '../../test-utils/PrettyTest';
const prettyTest = new PrettyTest([
    "misc/pg_catalog-timestamp-etc-1.sql",
    "misc/pg_catalog-timestamp-etc-2.sql",
    "misc/pg_catalog-timestamp-etc-3.sql",
    "misc/pg_catalog-timestamp-etc-4.sql",
    "misc/pg_catalog-timestamp-etc-5.sql",
    "misc/pg_catalog-timestamp-etc-6.sql",
    "misc/pg_catalog-timestamp-etc-7.sql",
    "misc/pg_catalog-timestamp-etc-8.sql",
    "misc/pg_catalog-timestamp-etc-9.sql",
    "misc/pg_catalog-timestamp-etc-10.sql",
    "misc/pg_catalog-timestamp-etc-11.sql",
    "misc/pg_catalog-timestamp-etc-12.sql",
    "misc/pg_catalog-timestamp-etc-13.sql",
    ///
    "misc/pg_catalog-1.sql",
    "misc/pg_catalog-2.sql",
    "misc/pg_catalog-3.sql",
    "misc/pg_catalog-4.sql",
    "misc/pg_catalog-5.sql",
    "misc/pg_catalog-6.sql",
    "misc/pg_catalog-7.sql",
    "misc/pg_catalog-8.sql",
    "misc/pg_catalog-9.sql",
    "misc/pg_catalog-10.sql",
    "misc/pg_catalog-11.sql",
    "misc/pg_catalog-12.sql",
    "misc/pg_catalog-13.sql",
    "misc/pg_catalog-14.sql",
    "misc/pg_catalog-15.sql",
    "misc/pg_catalog-16.sql",
    "misc/pg_catalog-17.sql"
]);

prettyTest.generateTests();
