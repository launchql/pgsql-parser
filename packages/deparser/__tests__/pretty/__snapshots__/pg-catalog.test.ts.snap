// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`non-pretty: misc/pg_catalog-1.sql 1`] = `"SELECT pg_catalog.json_object('{}')"`;

exports[`non-pretty: misc/pg_catalog-2.sql 1`] = `"SELECT * FROM generate_series(1, 5)"`;

exports[`non-pretty: misc/pg_catalog-3.sql 1`] = `"SELECT get_byte('\\xDEADBEEF'::bytea, 1)"`;

exports[`non-pretty: misc/pg_catalog-4.sql 1`] = `"SELECT now()"`;

exports[`non-pretty: misc/pg_catalog-5.sql 1`] = `"SELECT clock_timestamp()"`;

exports[`non-pretty: misc/pg_catalog-6.sql 1`] = `"SELECT to_char(now(), 'YYYY-MM-DD HH24:MI:SS')"`;

exports[`non-pretty: misc/pg_catalog-7.sql 1`] = `"SELECT json_build_object('name', 'Alice', 'age', 30)"`;

exports[`non-pretty: misc/pg_catalog-8.sql 1`] = `"SELECT pg_typeof(42), pg_typeof('hello'), pg_typeof(now())"`;

exports[`non-pretty: misc/pg_catalog-9.sql 1`] = `"SELECT SUBSTRING('abcdefg' FROM 2 FOR 3)"`;

exports[`non-pretty: misc/pg_catalog-10.sql 1`] = `"SELECT replace('hello world', 'l', 'L')"`;

exports[`non-pretty: misc/pg_catalog-11.sql 1`] = `"SELECT length('yolo')"`;

exports[`non-pretty: misc/pg_catalog-12.sql 1`] = `"SELECT POSITION('G' IN 'ChatGPT')"`;

exports[`non-pretty: misc/pg_catalog-13.sql 1`] = `"SELECT TRIM(BOTH FROM '  padded text  ')"`;

exports[`non-pretty: misc/pg_catalog-14.sql 1`] = `"SELECT ltrim('---abc', '-')"`;

exports[`non-pretty: misc/pg_catalog-15.sql 1`] = `"SELECT array_agg(id) FROM ( VALUES (1), (2), (3) ) AS t(id)"`;

exports[`non-pretty: misc/pg_catalog-16.sql 1`] = `"SELECT string_agg(name, ', ') FROM ( VALUES ('Alice'), ('Bob'), ('Carol') ) AS t(name)"`;

exports[`non-pretty: misc/pg_catalog-17.sql 1`] = `"SELECT json_agg(name) FROM ( VALUES ('A'), ('B') ) AS t(name)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-1.sql 1`] = `"CREATE TABLE public."ACTIVITY" ("ACTIVITY_ID" bigint NOT NULL, "PUBLISHED" timestamp(2) with time zone)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-2.sql 1`] = `"CREATE TABLE public."ACTIVITY" ("ACTIVITY_ID" bigint NOT NULL, "PUBLISHED" pg_catalog.timestamp(2))"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-3.sql 1`] = `"CREATE TABLE public."EXAMPLE" ("MESSAGE" text)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-4.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIME" ("START_TIME" pg_catalog.time)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-5.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIMETZ" ("START_TIME" time with time zone)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-6.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIME_PRECISION" ("START_TIME" pg_catalog.time(3))"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-7.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIMETZ_PRECISION" ("START_TIME" time(3) with time zone)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-8.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIMESTAMP" ("CREATED_AT" pg_catalog.timestamp)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-9.sql 1`] = `"CREATE TABLE public."EXAMPLE_TIMESTAMPTZ" ("CREATED_AT" timestamp with time zone)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-10.sql 1`] = `"CREATE TABLE public."EXAMPLE_INTERVAL" ("DURATION" interval)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-11.sql 1`] = `"CREATE TABLE public."EXAMPLE_INTERVAL_PRECISION" ("DURATION" interval(2))"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-12.sql 1`] = `"CREATE TABLE public."EXAMPLE_DATE" ("BIRTHDAY" date)"`;

exports[`non-pretty: misc/pg_catalog-timestamp-etc-13.sql 1`] = `"CREATE TABLE public."EXAMPLE_BOOL" ("IS_ACTIVE" boolean)"`;

exports[`pretty: misc/pg_catalog-1.sql 1`] = `"SELECT pg_catalog.json_object('{}')"`;

exports[`pretty: misc/pg_catalog-2.sql 1`] = `
"SELECT *
FROM generate_series(1, 5)"
`;

exports[`pretty: misc/pg_catalog-3.sql 1`] = `"SELECT get_byte('\\xDEADBEEF'::bytea, 1)"`;

exports[`pretty: misc/pg_catalog-4.sql 1`] = `"SELECT now()"`;

exports[`pretty: misc/pg_catalog-5.sql 1`] = `"SELECT clock_timestamp()"`;

exports[`pretty: misc/pg_catalog-6.sql 1`] = `"SELECT to_char(now(), 'YYYY-MM-DD HH24:MI:SS')"`;

exports[`pretty: misc/pg_catalog-7.sql 1`] = `"SELECT json_build_object('name', 'Alice', 'age', 30)"`;

exports[`pretty: misc/pg_catalog-8.sql 1`] = `
"SELECT
  pg_typeof(42),
  pg_typeof('hello'),
  pg_typeof(now())"
`;

exports[`pretty: misc/pg_catalog-9.sql 1`] = `"SELECT SUBSTRING('abcdefg' FROM 2 FOR 3)"`;

exports[`pretty: misc/pg_catalog-10.sql 1`] = `"SELECT replace('hello world', 'l', 'L')"`;

exports[`pretty: misc/pg_catalog-11.sql 1`] = `"SELECT length('yolo')"`;

exports[`pretty: misc/pg_catalog-12.sql 1`] = `"SELECT POSITION('G' IN 'ChatGPT')"`;

exports[`pretty: misc/pg_catalog-13.sql 1`] = `"SELECT TRIM(BOTH FROM '  padded text  ')"`;

exports[`pretty: misc/pg_catalog-14.sql 1`] = `"SELECT ltrim('---abc', '-')"`;

exports[`pretty: misc/pg_catalog-15.sql 1`] = `
"SELECT array_agg(id)
FROM ( VALUES
  (1),
  (2),
  (3) ) AS t(id)"
`;

exports[`pretty: misc/pg_catalog-16.sql 1`] = `
"SELECT string_agg(name, ', ')
FROM ( VALUES
  ('Alice'),
  ('Bob'),
  ('Carol') ) AS t(name)"
`;

exports[`pretty: misc/pg_catalog-17.sql 1`] = `
"SELECT json_agg(name)
FROM ( VALUES
  ('A'),
  ('B') ) AS t(name)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-1.sql 1`] = `
"CREATE TABLE public."ACTIVITY" (
  "ACTIVITY_ID" bigint NOT NULL,
  "PUBLISHED" timestamp(2) with time zone
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-2.sql 1`] = `
"CREATE TABLE public."ACTIVITY" (
  "ACTIVITY_ID" bigint NOT NULL,
  "PUBLISHED" pg_catalog.timestamp(2)
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-3.sql 1`] = `
"CREATE TABLE public."EXAMPLE" (
  "MESSAGE" text
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-4.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIME" (
  "START_TIME" pg_catalog.time
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-5.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIMETZ" (
  "START_TIME" time with time zone
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-6.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIME_PRECISION" (
  "START_TIME" pg_catalog.time(3)
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-7.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIMETZ_PRECISION" (
  "START_TIME" time(3) with time zone
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-8.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIMESTAMP" (
  "CREATED_AT" pg_catalog.timestamp
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-9.sql 1`] = `
"CREATE TABLE public."EXAMPLE_TIMESTAMPTZ" (
  "CREATED_AT" timestamp with time zone
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-10.sql 1`] = `
"CREATE TABLE public."EXAMPLE_INTERVAL" (
  "DURATION" interval
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-11.sql 1`] = `
"CREATE TABLE public."EXAMPLE_INTERVAL_PRECISION" (
  "DURATION" interval(2)
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-12.sql 1`] = `
"CREATE TABLE public."EXAMPLE_DATE" (
  "BIRTHDAY" date
)"
`;

exports[`pretty: misc/pg_catalog-timestamp-etc-13.sql 1`] = `
"CREATE TABLE public."EXAMPLE_BOOL" (
  "IS_ACTIVE" boolean
)"
`;
